dependencies/spirv/CMakeLists.txt

```
@@ -0,0 +1,37 @@
+include(FetchContent)
+include(CMakeDependentOption)
+
+cmake_dependent_option(FETCH_SPIRV_HEADERS "Fetch SPIR-V headers from remote repo (disable and use SPIRV-Headers_SOURCE_DIR to use an existing checkout)." ON TARGET_SPIRV ON)
+
+if (TARGET_SPIRV)
+    if (DEFINED SPIRV-Headers_SOURCE_DIR)
+        if (IS_DIRECTORY ${SPIRV-Headers_SOURCE_DIR})
+            set(HALIDE_SPIRV_SOURCE_DIR ${SPIRV-Headers_SOURCE_DIR})
+            set(HALIDE_SPIRV_INCLUDE_DIR ${SPIRV-Headers_SOURCE_DIR}/include)    
+        else()
+            message(FATAL_ERROR "SPIRV-Headers were not found - please specify a valid path with 'SPIRV-Headers_SOURCE_DIR'!")
+        endif()
+    elseif(FETCH_SPIRV_HEADERS)
+        set(SPIRV_VERSION 1.5.4)
+        message(STATUS "Fetching SPIR-V ${SPIRV_VERSION}...")
+        FetchContent_Declare(SPIRV-Headers
+                            GIT_REPOSITORY https://github.com/KhronosGroup/SPIRV-Headers.git
+                            GIT_TAG ${SPIRV_VERSION}
+                            GIT_SHALLOW TRUE
+                            GIT_SUBMODULES "")
+
+        FetchContent_MakeAvailable(SPIRV-Headers)
+
+        set(HALIDE_SPIRV_SOURCE_DIR ${SPIRV-Headers_SOURCE_DIR})
+        set(HALIDE_SPIRV_INCLUDE_DIR ${SPIRV-Headers_SOURCE_DIR}/include)    
+    endif()
+
+    if (DEFINED HALIDE_SPIRV_INCLUDE_DIR)
+        add_library(Halide_SPIRV INTERFACE)
+        target_include_directories(Halide_SPIRV SYSTEM INTERFACE ${HALIDE_SPIRV_INCLUDE_DIR})
+        message(STATUS "Using SPIR-V headers from ${HALIDE_SPIRV_INCLUDE_DIR}...")
+    else()
+        message(FATAL_ERROR "SPIRV-Headers were not found - please enable FETCH_SPIRV_HEADERS or specify a valid path with 'SPIRV-Headers_SOURCE_DIR'!")
+    endif()
+
+endif()
```

Here is a correct implementation:

```suggestion
cmake_dependent_option(
    FETCH_SPIRV_HEADERS "Fetch SPIR-V headers from remote repo (disable and use SPIRV-Headers_SOURCE_DIR to use an existing checkout)." OFF
    TARGET_SPIRV OFF
)

if (FETCH_SPIRV_HEADERS)
    set(SPIRV_VERSION 1.5.4)
    message(STATUS "Fetching SPIR-V ${SPIRV_VERSION}...")
    FetchContent_Declare(
        SPIRV-Headers
        GIT_REPOSITORY https://github.com/KhronosGroup/SPIRV-Headers.git
        GIT_TAG ${SPIRV_VERSION}
        GIT_SHALLOW TRUE
        GIT_SUBMODULES ""
    )

    FetchContent_MakeAvailable(SPIRV-Headers)

    set(SPIRV_INCLUDE_DIR ${SPIRV-Headers_SOURCE_DIR}/include)    
else ()
    find_path(SPIRV_INCLUDE_DIR spirv/1.0/spirv.h)
endif ()

if (NOT SPIRV_INCLUDE_DIR)
    message(FATAL_ERROR "SPIRV-Headers were not found - please enable FETCH_SPIRV_HEADERS or specify a valid path with 'SPIRV_INCLUDE_DIR'!")
endif ()

message(STATUS "Using SPIR-V headers from ${SPIRV_INCLUDE_DIR}...")

add_library(Halide_SPIRV INTERFACE IMPORTED)
target_include_directories(Halide_SPIRV INTERFACE "${SPIRV_INCLUDE_DIR}")
```

---

I would prefer to move Halide away from using FetchContent for anything by default. It's okay (until we reach CMake 3.24) to include it as an opt-in convenience, but we've already had issues raised with people trying to build us in air-gapped environments. Better to get an error message we control than one that some internal CMake module controls.

In the case when FetchContent is not used, we should at least _try_ to look for the header we're using. That's the purpose of the `find_path` command.

Int he case when FetchContent is used, since we are not using their CMake targets (and are instead making our own), we shouldn't use `MakeAvailable` since that will run all of their CMake code inside our project. Instead, we'll just `Populate` it to trigger the download, but nothing else.